### PR TITLE
Pod shutdown tests porting

### PR DESCRIPTION
**Pull request by sankarpn**
The tests from JUnit4 integration-tests/src/test/java/oracle/kubernetes/operator/ItPodsShutdown.java test class is ported to JUnit5

The tests covers the following scenarios

- Set Shutdown property values SHUTDOWN_IGNORE_SESSIONS, SHUTDOWN_TYPE, SHUTDOWN_TIMEOUT at all levels such as admin server, cluster and independent managed server levels and verifies the values are applied as per the documented feature

- Set Shutdown property values SHUTDOWN_IGNORE_SESSIONS, SHUTDOWN_TYPE, SHUTDOWN_TIMEOUT at all levels such as admin server, cluster and independent managed server levels and overrides the SHUTDOWN_TYPE as Env item and verifies the values are applied as per the documented feature and overridden SHUTDOWN_TYPE at env is applied to every server than the one set at server/cluster since env item has high precedence.

Test Results
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/1961
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/1960
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/1956
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/1955

### CODE REVIEW

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 1]
```diff
@@ -156,15 +118,158 @@ private static void createAndVerifyMiiDomain() {
...
+   *<p>This use case shows how to add shutdown options at domain and how to override them using ENV variable.
+   *
+   * @throws ApiException when getting log fails
+   */
+  @Test
+  @DisplayName("Verify shutdown rules when shutdown properties are defined at different levels ")
+  public void testShutdownPropsEnvOverride() throws ApiException {
+
+
+    // create Shitdown objects for each server and cluster
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:237]: shutdown

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 2]
```diff
@@ -183,108 +288,117 @@ private static void createAndVerifyMiiDomain() {
...
         domainUid, domainNamespace, miiImage);
     createDomainAndVerify(domain, domainNamespace);
 
-    // check that admin server pod exists in the domain namespace
-    logger.info("Checking that admin server pod {0} exists in namespace {1}",
+    // check that admin service exists in the domain namespace
+    logger.info("Checking that admin service {0} exists in namespace {1}",
         adminServerPodName, domainNamespace);
-    checkPodExists(adminServerPodName, domainUid, domainNamespace);
+    checkServiceExists(adminServerPodName, domainNamespace);
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:338]: you can call checkPodReadyAndServiceExists(String podName, String domainUid, String namespace) method

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 3]
```diff
@@ -183,108 +288,117 @@ private static void createAndVerifyMiiDomain() {
...
-      String managedServerPodName = managedServerPrefix + i;
+      String managedServerPodName = managedServerPodNamePrefix + i;
 
-      // check that the managed server pod exists in the domain namespace
-      logger.info("Checking that managed server pod {0} exists in namespace {1}",
+      // check that the managed server service exists in the domain namespace
+      logger.info("Checking that managed server service {0} exists in namespace {1}",
           managedServerPodName, domainNamespace);
-      checkPodExists(managedServerPodName, domainUid, domainNamespace);
+      checkServiceExists(managedServerPodName, domainNamespace);
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:351]: you can call checkPodReadyAndServiceExists(String podName, String domainUid, String namespace) method

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 4]
```diff
@@ -183,108 +288,117 @@ private static void createAndVerifyMiiDomain() {
...
+    }
 
+    // check for independent managed server pods existence in the domain namespace
+    for (String podName : new String[]{indManagedServerPodName2}) {
       // check that the managed server service exists in the domain namespace
       logger.info("Checking that managed server service {0} exists in namespace {1}",
-          managedServerPodName, domainNamespace);
-      checkServiceExists(managedServerPodName, domainNamespace);
+          podName, domainNamespace);
+      checkServiceExists(podName, domainNamespace);
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:364]: you can call checkPodReadyAndServiceExists(String podName, String domainUid, String namespace) method

**Review by vanajamukkara — CHANGES_REQUESTED** [review]

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 5]
```diff
@@ -156,15 +118,127 @@ private static void createAndVerifyMiiDomain() {
...
+    checkPodDoesNotExist(managedServerPodNamePrefix + 2, domainUid, domainNamespace);
+    checkPodDoesNotExist(indManagedServerPodName1, domainUid, domainNamespace);
+    checkPodDoesNotExist(indManagedServerPodName2, domainUid, domainNamespace);
+  }
+
+  /**
+   * Add shutdown properties at all levels and verify.
+   * 1. Creates shutdown object for admin server, cluster and managed servers.
+   * 2. Creates domain resource with the above shutdown objects.
+   * 3. After all the pods are started, verifies the server.out for each server to see whether the nodemanager env
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: Fixed the javadoc

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 6]
```diff
@@ -188,103 +262,112 @@ private static void createAndVerifyMiiDomain() {
...
+      checkServiceExists(managedServerPodName, domainNamespace);
 
       // check that the managed server pod is ready
       logger.info("Checking that managed server pod {0} is ready in namespace {1}",
           managedServerPodName, domainNamespace);
       checkPodReady(managedServerPodName, domainUid, domainNamespace);
+    }
 
+    // check for independent managed server pods existence in the domain namespace
+    //TODO - only one of the independent managed server is coming up, bug???
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: Fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 7]
```diff
@@ -156,15 +118,158 @@ private static void createAndVerifyMiiDomain() {
...
+   *
+   *<p>This use case shows how to add shutdown options at domain and how to override them at server level.
+   * @throws ApiException when getting log fails
+   */
+  @Test
+  @DisplayName("Verify shutdown rules when shutdown properties are defined at different levels ")
+  public void testShutdownPropsAllLevels() throws ApiException {
+
+
+    // create Shitdown objects for each server and cluster
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 8]
```diff
@@ -94,58 +106,8 @@ public static void initAll(@Namespaces(2) List<String> namespaces) {
...
-   * Create a model in image domain and verify the server pods are ready.
-   */
-  private static void createAndVerifyMiiDomain() {
-
     // get the pre-built image created by IntegrationTestWatcher
-    String miiImage = MII_BASIC_IMAGE_NAME + ":" + MII_BASIC_IMAGE_TAG;
+    miiImage = MII_BASIC_IMAGE_NAME + ":" + MII_BASIC_IMAGE_TAG;
 
     // docker login and push image to docker registry if necessary
     dockerLoginAndPushImageToRegistry(miiImage);
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: removed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 9]
```diff
@@ -183,108 +288,117 @@ private static void createAndVerifyMiiDomain() {
...
     }
-    return found;
+  }
+
+  // Crate a ConfigMap with a model to add a 2 independent managed servers
+  private static void createModelConfigMap(String configMapName, String model) {
+    Map<String, String> labels = new HashMap<>();
+    labels.put("weblogic.domainUid", domainUid);
+    Map<String, String> data = new HashMap<>();
+    data.put("configured-cluster.yaml", model);
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 10]
```diff
@@ -183,108 +288,117 @@ private static void createAndVerifyMiiDomain() {
...
-    }
-    if (props.length == propFound.size()) {
-      found = true;
+
+  // get pod log which includes the server.out logs and verify the messages contain the set shutdown properties
+  private void verifyServerLog(String namespace, String podName, String[] envVars) throws ApiException {
+    String podLog = TestActions.getPodLog(podName, namespace);
+    for (String envVar : envVars) {
+      logger.info("Checking Pod {0} for server startup property {1}", podName, envVar);
+      assertTrue(podLog.contains(envVar));
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 11]
```diff
@@ -156,15 +118,158 @@ private static void createAndVerifyMiiDomain() {
...
+   *
+   *<p>When server pods starts up the server.out log will show the options applied to the servers. The test verifies
+   * the logs and determine the outcome of the test.
+   *
+   *<p>This use case shows how to add shutdown options at domain and how to override them at server level.
+   * @throws ApiException when getting log fails
+   */
+  @Test
+  @DisplayName("Verify shutdown rules when shutdown properties are defined at different levels ")
+  public void testShutdownPropsAllLevels() throws ApiException {
```
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:173]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java (hunk 1 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: fixed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java (hunk 2 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: modified

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java (hunk 3 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: modified

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java (hunk 4 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: modified

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (vanajamukkara) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:173]: I don't see any change for this?

[the remark this answers is not in the document]
Comment (vanajamukkara) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:173]: you missed deleting throws from test method..

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java [hunk 12]
```diff
@@ -94,77 +104,167 @@ public static void initAll(@Namespaces(2) List<String> namespaces) {
...
+   *<p>When server pods starts up the server.out log will show the shutdown options applied to the servers.
+   * The test verifies the logs and determine the outcome of the test.
+   *
+   *<p>This use case shows how to add shutdown options at domain and how to override them using ENV variable.
+   *
+   * @throws ApiException when getting log fails
+   */
+  @Test
+  @DisplayName("Verify shutdown rules when shutdown properties are defined at different levels ")
+  public void testShutdownPropsEnvOverride() throws ApiException {
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java:229]: remove throws ApiException

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: yeah, fixed it.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java (hunk 12 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItPodsShutdownOption.java]: fixed

**Review by vanajamukkara — APPROVED** [review]
LGTM

**Review by anpanigr — APPROVED** [review]
LGTM

**Pull request merged**